### PR TITLE
Changed JP2 map to figure test

### PR DESCRIPTION
**Pull request by jeffreypaul15**
### Description

Converted the test for JP2 map to a figure test.
Does this require a changelog?

### CODE REVIEW

File: sunpy/net/tests/test_helioviewer.py [hunk 1]
```diff
@@ -88,16 +89,16 @@ def test_get_jp2_header(self, client):
...
     def test_download_jp2_map(self, client):
         """
         Tests getJP2Image API method with Map.
         """
-        # TODO: make this a figure test.
         filepath = client.download_jp2('2012/01/01', observatory='SOHO',
                                        instrument='MDI', measurement='continuum')
         map_ = sunpy.map.Map(filepath)
-        assert isinstance(map_, sunpy.map.GenericMap)
-        os.remove(filepath)
```
Comment (nabobalis) [inline sunpy/net/tests/test_helioviewer.py:100]: Should we still not delete the file or at least download it into a temporary directory.

File: sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json [hunk 2]
```diff
@@ -45,6 +46,9 @@
   "sunpy.timeseries.tests.test_timeseriesbase.test_noaa_txt_pre_peek": "fbda770f164bd38a7741d3393e9f93040b188e928521cc53379ba2b78beb582c",
   "sunpy.timeseries.tests.test_timeseriesbase.test_generic_ts_peek": "73a71e717b7dd0dd345f0f8a8bfa7f06387607417b3ecd123adfd8f630f93d51",
   "sunpy.visualization.animator.tests.test_basefuncanimator.test_lineanimator_figure": "dbf47ab983a844621f912b2f5a2d9ee657d3f661678c16a07afc7217f50e3fd4",
+  "sunpy.visualization.animator.tests.test_mapsequence.test_map_sequence_animator_wcs_simple_plot": "0b4d489854b85b6fe469c61c8b1920183e06bb59f2075630d23f46a3c516e030",
+  "sunpy.visualization.animator.tests.test_mapsequence.test_map_sequence_animator_wcs_colorbar_buttons": "fcd5d64e48e45c3eebe73a835a22921bfb80b5a8b63eb85faddb68a2cfb447d1",
+  "sunpy.visualization.animator.tests.test_mapsequence.test_map_sequence_animator_wcs_colorbar_buttons_default_labels": "efc609d0263597f1c66ff578262f14a2e1cee73bf90e2fb2f44334d82fed85f6",
```
Comment (nabobalis) [inline sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json:51]:
```suggestion
```

Comment (jeffreypaul15) [reply sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json:51]: Sorry this keeps happening, I need to fix this.

File: sunpy/net/tests/test_helioviewer.py (hunk 1 above)
Comment (jeffreypaul15) [reply sunpy/net/tests/test_helioviewer.py:100]: Oh yeah, shouldn't have removed that line.

**Comment by jeffreypaul15** [comment]
@nabobalis https://github.com/sunpy/sunpy/blob/7a7585476217fd3952c91d3bbf5266f73a60bb7e/sunpy/map/mapbase.py#L1142
Is a small test that isn't covered according to codecov, can I add it to this PR itself?

**Comment by jeffreypaul15** [comment]
Why are the figure tests failing, has something gone wrong?

**Comment by nabobalis** [comment]
The 37 dev build has been broken by some sort of change upstream. The normal one be ok.

**Comment by jeffreypaul15** [comment]
> The 37 dev build has been broken by some sort of change upstream. The normal one be ok.

So nothing wrong with the code here?

**Comment by nabobalis** [comment]
> > The 37 dev build has been broken by some sort of change upstream. The normal one be ok.
> 
> So nothing wrong with the code here?

Both the figure tests failed. So there is something wrong here as well.

**Comment by nabobalis** [comment]
A fix for the 37 figure build was merged into master.

**Comment by jeffreypaul15** [comment]
Hmm this is odd, my tests are passing locally..

**Comment by nabobalis** [comment]
That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.

**Comment by jeffreypaul15** [comment]
> That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.
Yeah.. Is there any other way to get the figure hash?

**Comment by nabobalis** [comment]
> > That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.
> 
> Yeah.. Is there any other way to get the figure hash?

Other than running the tox environment? No. Someone else could run it for you.

**Comment by nabobalis** [comment]
> > > That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.
> > 
> > 
> > Yeah.. Is there any other way to get the figure hash?
> 
> Other than running the tox environment? No. Someone else could run it for you.

I've ran it locally and will push the changed hash.

**Review by nabobalis — APPROVED** [review]

File: sunpy/net/tests/test_helioviewer.py [hunk 3]
```diff
@@ -88,17 +89,20 @@ def test_get_jp2_header(self, client):
         assert ('fits' in header1.keys()) and ('fits' in header2.keys())
 
     @skip_glymur
+    @figure_test
     def test_download_jp2_map(self, client):
         """
-        Tests getJP2Image API method with Map.
+        Tests getJP2Image API method with Map with a figure test.
         """
         # TODO: make this a figure test.
```
Comment (ayshih) [inline sunpy/net/tests/test_helioviewer.py:97]:
```suggestion
```

File: sunpy/net/tests/test_helioviewer.py [hunk 4]
```diff
@@ -88,17 +89,20 @@ def test_get_jp2_header(self, client):
...
         filepath = client.download_jp2('2012/01/01', observatory='SOHO',
                                        instrument='MDI', measurement='continuum')
```
Comment (ayshih) [inline sunpy/net/tests/test_helioviewer.py:99]: This isn't your doing, but this test should really be downloading to a `tempfile.TemporaryDirectory()`, with a call to `.cleanup()` at the end of the test.

File: setup.cfg [hunk 5]
```diff
@@ -78,6 +78,7 @@ tests =
   pytest-mock
   pytest-mpl>=0.12 # First version to support our figure tests
   tox
+  tox-pypi-filter
```
Comment (ayshih) [inline setup.cfg:81]: Why is this needed?

**Comment by jeffreypaul15** [comment]
> > > > That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.
> > > 
> > > 
> > > Yeah.. Is there any other way to get the figure hash?
> > 
> > 
> > Other than running the tox environment? No. Someone else could run it for you.
> 
> I've ran it locally and will push the changed hash. Hopefully it should pass now.

Thanks! @nabobalis

File: sunpy/net/tests/test_helioviewer.py (hunk 4 above)
Comment (jeffreypaul15) [reply sunpy/net/tests/test_helioviewer.py:99]: Alright, I shall make that change.

File: setup.cfg (hunk 5 above)
Comment (nabobalis) [reply setup.cfg:81]: I added this since we don't actually pre-install it. Tox never uses the test pin files.

**Comment by nabobalis** [comment]
> > > > > That is odd, its trying to download the figure which means the hashes don't seem to line up with what circleci created.
> > > > 
> > > > 
> > > > Yeah.. Is there any other way to get the figure hash?
> > > 
> > > 
> > > Other than running the tox environment? No. Someone else could run it for you.
> > 
> > 
> > I've ran it locally and will push the changed hash. Hopefully it should pass now.
> 
> The py37 test is still failing.
> Also, what are the commands used to generate the hashes? I've been using wrong commands all this while.

They are in the dev docs: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html

File: sunpy/net/tests/test_helioviewer.py [hunk 6]
```diff
@@ -88,16 +90,20 @@ def test_get_jp2_header(self, client):
...
         """
-        # TODO: make this a figure test.
-        filepath = client.download_jp2('2012/01/01', observatory='SOHO',
-                                       instrument='MDI', measurement='continuum')
-        map_ = sunpy.map.Map(filepath)
-        assert isinstance(map_, sunpy.map.GenericMap)
-        os.remove(filepath)
+        filepath = tempfile.TemporaryDirectory()
+        client.download_jp2('2012/01/01', observatory='SOHO',
+                            instrument='MDI', measurement='continuum', directory=tempfile.name)
```
Comment (nabobalis) [inline sunpy/net/tests/test_helioviewer.py:100]: Could we not use TemporaryDirectory as a context manager? 

In fact we could use pytests tmpdir instead and that should auto clean up at the end of the test.

Comment (jeffreypaul15) [reply sunpy/net/tests/test_helioviewer.py:100]: Okay, shall I make this change?

**Comment by jeffreypaul15** [comment]
> They are in the dev docs: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html

Shouldn't this command be updated in the docs as well?
`tox -e py38-figure -- --mpl-generate-hash-library=sunpy/tests/figure_hashes_mpl_332_ft_261_astropy_403.json`

File: sunpy/net/tests/test_helioviewer.py (hunk 6 above)
Comment (nabobalis) [reply sunpy/net/tests/test_helioviewer.py:100]: Sure.

**Comment by nabobalis** [comment]
> > They are in the dev docs: [docs.sunpy.org/en/latest/dev_guide/contents/tests.html](https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html)
> 
> Shouldn't this command be updated in the docs as well?
> `tox -e py38-figure -- --mpl-generate-hash-library=sunpy/tests/figure_hashes_mpl_332_ft_261_astropy_403.json`

You mean the filename? I guess.

File: setup.cfg (hunk 5 above)
Comment (Cadair) [reply setup.cfg:81]: We could copy astropy and use to to control the env in which tox runs: https://github.com/astropy/astropy/blob/master/tox.ini#L10

This does expose us to some edge cases, but in general should work fine.

Comment (nabobalis) [reply setup.cfg:81]: What kind of edge cases?

Comment (nabobalis) [reply setup.cfg:81]: I'll make this change in another PR.

Comment (nabobalis) [inline setup.cfg:81]:
```suggestion
```

**Comment by nabobalis** [comment]
The hash on the "py37-figure-devdeps" failed on the last go around.

**Comment by jeffreypaul15** [comment]
> The hash on the "py37-figure-devdeps" failed on the last go around.

Yeah, I'm not sure why this is happening.

**Comment by nabobalis** [comment]
> > The hash on the "py37-figure-devdeps" failed on the last go around.
> 
> Yeah, I'm not sure why this is happening.

The hash is wrong I would assume.

**Comment by jeffreypaul15** [comment]
> > > The hash on the "py37-figure-devdeps" failed on the last go around.
> > 
> > 
> > Yeah, I'm not sure why this is happening.
> 
> The hash is wrong I would assume.

This seems to be an issue as it's passing locally, and I haven't made any changes to the hashes you committed

File: sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json [hunk 7]
```diff
@@ -32,6 +32,7 @@
   "sunpy.map.tests.test_plotting.test_heliographic_quadrangle_width_height": "f1dc10238c83fbdf510d1f912c921fe3b53f437f9a46f3b71ad46b48273369fb",
   "sunpy.map.tests.test_plotting.test_heliographic_quadrangle_top_right": "b4d2229be76ed1fdcfa5740aaeb26ab6eed92608661cf2f34bca3685f524caf1",
   "sunpy.map.tests.test_plotting.test_heliographic_grid_annotations": "eae3267f21837571a7b10bf1e2fe0f672cf59c11ebefa4e5eadc3e9906609be3",
+  "sunpy.net.tests.test_helioviewer.test_download_jp2_map": "438fdcccc6f879491c876ce24a591843bb60d724e354d1c8124678a356ff3426",
```
Comment (nabobalis) [inline sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json:35]:
```suggestion
  "sunpy.net.tests.test_helioviewer.test_download_jp2_map": "67b17ac7c343a1d568d9dfa2bb12e0041e2fc53e8ca1c0e1c2c7693eda2bd0d8",
```
This is what I got locally.

Comment (jeffreypaul15) [reply sunpy/tests/figure_hashes_mpl_dev_ft_261_astropy_dev.json]: My apologies, might've mixed it up.

File: sunpy/net/tests/test_helioviewer.py [hunk 8]
```diff
@@ -88,16 +89,17 @@ def test_get_jp2_header(self, client):
         assert ('fits' in header1.keys()) and ('fits' in header2.keys())
 
     @skip_glymur
-    def test_download_jp2_map(self, client):
+    @figure_test
+    def test_download_jp2_map(self, client, tmpdir):
```
Comment (ayshih) [inline sunpy/net/tests/test_helioviewer.py:93]:
```suggestion
    def test_download_jp2_map(self, client, tmp_path):
```

File: sunpy/net/tests/test_helioviewer.py [hunk 9]
```diff
@@ -88,16 +89,17 @@ def test_get_jp2_header(self, client):
...
         """
-        Tests getJP2Image API method with Map.
+        Tests getJP2Image API method with Map with a figure test.
         """
-        # TODO: make this a figure test.
         filepath = client.download_jp2('2012/01/01', observatory='SOHO',
-                                       instrument='MDI', measurement='continuum')
+                                       instrument='MDI', measurement='continuum',
+                                       directory=tmpdir.dirname)
```
Comment (ayshih) [inline sunpy/net/tests/test_helioviewer.py:99]:
```suggestion
                                       directory=tmp_path)
```

File: sunpy/net/tests/test_helioviewer.py [hunk 10]
```diff
@@ -88,16 +89,17 @@ def test_get_jp2_header(self, client):
...
         map_ = sunpy.map.Map(filepath)
-        assert isinstance(map_, sunpy.map.GenericMap)
-        os.remove(filepath)
+        with pytest.raises(SunpyMetadataWarning, match="Missing metadata for observer: assuming Earth-based observer."):
+            map_.peek()
```
Comment (ayshih) [inline sunpy/net/tests/test_helioviewer.py:102]: Probably shouldn't use `peek()` to generate the figure, but we don't need to open that can of worms here

**Review by ayshih — APPROVED** [review]
Good call on using a pytest fixture, but I think you should use `tmp_path` instead of `tmpdir`

**Comment by nabobalis** [comment]
Thanks again @jeffreypaul15
